### ISSUE TITLE
Optimize train_test_split performance

### ISSUE THREAD

**Issue by TomAugspurger** [dask/dask-ml#174]
Followup to https://github.com/dask/dask-ml/pull/172

With that PR, `train_test_split` internally uses `ShuffleSplit` to generate the dask array containing the indices to slice. It's possible that the index array isn't on the same machine as the data array.

![mydask](https://user-images.githubusercontent.com/1312546/40365395-12b6a596-5d9a-11e8-88a3-91eca440d9aa.png)

Each of the A-like branches has the potential for communication.

I think the solution is to generate the indexes and slice in the same task.
